web_src/js/components/RepoBranchTagSelector.vue

```diff
@@ -0,0 +1,293 @@
+      return items;
+    },
+    showNoResults() {
+      return this.filteredItems.length === 0 && !this.showCreateNewBranch;
+    },
+    showCreateNewBranch() {
+      if (this.disableCreateBranch || !this.searchTerm) {
+        return false;
+      }
+      return this.items.filter((item) => item.name.toLowerCase() === this.searchTerm.toLowerCase()).length === 0;
+    },
+    formActionUrl() {
+      return `${this.repoLink}/branches/_new/${pathEscapeSegments(this.branchNameSubURL)}`;
+    },
+  },
+
+  watch: {
+    menuVisible(visible) {
+      if (visible) {
+        this.focusSearchField();
+      }
+    }
+  },
+
+  beforeMount() {
+    if (this.viewType === 'tree') {
+      this.isViewTree = true;
+      this.refNameText = this.commitIdShort;
+    } else if (this.viewType === 'tag') {
+      this.isViewTag = true;
+      this.refNameText = this.tagName;
+    } else {
+      this.isViewBranch = true;
+      this.refNameText = this.branchName;
+    }
+
+    document.body.addEventListener('click', (event) => {
+      if (this.$el.contains(event.target)) return;
+      if (this.menuVisible) {
+        this.menuVisible = false;
+      }
+    });
+  },
+
+  methods: {
+    selectItem(item) {
+      const prev = this.getSelected();
+      if (prev !== null) {
+        prev.selected = false;
+      }
+      item.selected = true;
+      const url = (item.tag) ? this.tagURLPrefix + item.url + this.tagURLSuffix : this.branchURLPrefix + item.url + this.branchURLSuffix;
+      if (!this.branchForm) {
+        window.location.href = url;
+      } else {
+        this.isViewTree = false;
+        this.isViewTag = false;
+        this.isViewBranch = false;
+        this.$refs.dropdownRefName.textContent = item.name;
+        if (this.setAction) {
+          $(`#${this.branchForm}`).attr('action', url);
+        } else {
+          $(`#${this.branchForm} input[name="refURL"]`).val(url);
+        }
+        $(`#${this.branchForm} input[name="ref"]`).val(item.name);
+        if (item.tag) {
+          this.isViewTag = true;
+          $(`#${this.branchForm} input[name="refType"]`).val('tag');
+        } else {
+          this.isViewBranch = true;
+          $(`#${this.branchForm} input[name="refType"]`).val('branch');
+        }
+        if (this.submitForm) {
+          $(`#${this.branchForm}`).trigger('submit');
+        }
+        this.menuVisible = false;
+      }
+    },
+    createNewBranch() {
+      if (!this.showCreateNewBranch) return;
+      $(this.$refs.newBranchForm).trigger('submit');
+    },
+    focusSearchField() {
+      nextTick(() => {
+        this.$refs.searchField.focus();
+      });
+    },
+    getSelected() {
+      for (let i = 0, j = this.items.length; i < j; ++i) {
+        if (this.items[i].selected) return this.items[i];
+      }
+      return null;
+    },
+    getSelectedIndexInFiltered() {
+      for (let i = 0, j = this.filteredItems.length; i < j; ++i) {
```

Comment:
I don't think we can cache the length to traverse through.
Assuming the list is mutated while we traverse through it, it would throw an index-out-of-bounds-exception, or not traverse through all elements. 
```suggestion
      for (let i = 0; i < this.filteredItems.length; ++i) {
```

Reply:
This PR only moves the code into Vue SFC, **doesn't touch the unrelated logic**